tests/python_package_test/test_basic.py

```diff
@@ -609,3 +609,44 @@ def test_custom_objective_safety():
     good_bst_multi.update(fobj=_good_gradients)
     with pytest.raises(ValueError, match=re.escape(f"number of models per one iteration ({nclass})")):
         bad_bst_multi.update(fobj=_bad_gradients)
+
+
+def test_training_with_pandas_nullable_dtypes():
+    pd = pytest.importorskip('pandas')
+    rand_ints = np.random.randint(0, 2, size=100)
+    unif_rands = np.random.rand(100)
+    df = pd.DataFrame(
+        {
+            'x1': rand_ints,
+            'x2': unif_rands,
+            'x3': unif_rands < 0.5,
+            'x4': pd.arrays.SparseArray(rand_ints),
+        }
+    )
+    # introduce some missing values
+    df.loc[1, 'x1'] = np.nan
+    df.loc[2, 'x2'] = np.nan
+    df.loc[3, 'x3'] = np.nan
+    # the previous line turns x3 into object dtype in recent versions of pandas
+    df['x3'] = df['x3'].astype(np.float64)
+    y = np.random.rand(100)
+
+    # train with regular dtypes
+    params = {'num_leaves': 15}
+    ds = lgb.Dataset(df, y)
+    bst = lgb.train(params, ds, num_boost_round=5)
+
+    # convert to nullable dtypes
+    df2 = df.copy()
+    df2['x1'] = df2['x1'].astype('Int32')
+    df2['x2'] = df2['x2'].astype('Float64')
+    df2['x3'] = df2['x3'].astype('boolean')
+
+    # test training succeeds
+    ds_nullable_dtypes = lgb.Dataset(df2, y)
+    bst_nullable_dtypes = lgb.train(params, ds_nullable_dtypes, num_boost_round=5)
+
+    # test equal predictions
+    preds = bst.predict(df)
+    preds_nullable_dtypes = bst_nullable_dtypes.predict(df2)
+    np.testing.assert_allclose(preds, preds_nullable_dtypes)
```

Comment:
I like the structure of this test, but I think it could pass even if this support wasn't working correctly. For example, with purely-random input data it's possible that both of the models trained here will just return the sample mean (e.g. not add any leaves, because all splits have 0 or too-low gain). In that situation, the types of issues you'd expect from a type-casting change (like loss of precision or incorrect data because of incorrect pointer math) might not be caught.

Could you modify this test to use one of the datasets used in the other tests here, and check things like the score on the training data being "ok" and the model having more leaves than `num_boost_round`?

Reply:
I've modified the test for the nullable dtypes in e1cf6c9cd759a735dcf5bf68f015e951b255c677. I used a synthetic dataset because I needed a boolean and some ints and most of the real datasets have floats but I made sure that all of the features were used and that the score was good. Let me know what you think.

Reply:
Ok yeah that looks good to me! Thanks